From 2ffbde6d613bd1678356f8cf7e26430145c01c93 Mon Sep 17 00:00:00 2001
From: Thiago Macieira <thiago.macieira@intel.com>
Date: Wed, 10 Oct 2018 19:53:40 -0700
Subject: [llvm] Detect Clear Linux and apply Clear's default linker options

Enum placed at the end to avoid Binary Compatibility break when
backporting this patch.

Signed-off-by: Thiago Macieira <thiago.macieira@intel.com>
---
 clang/include/clang/Driver/Distro.h   |  4 ++++
 clang/lib/Driver/Distro.cpp           | 13 +++++++++++++
 clang/lib/Driver/ToolChains/Linux.cpp | 14 +++++++++-----
 clang/unittests/Driver/DistroTest.cpp | 16 ++++++++++++++++
 4 files changed, 42 insertions(+), 5 deletions(-)

diff --git llvm-project/tools/clang/include/clang/Driver/Distro.h llvm-project/tools/clang/include/clang/Driver/Distro.h
index 038d4ce75d80..6253d9d56ef0 100644
--- llvm-project/tools/clang/include/clang/Driver/Distro.h
+++ llvm-project/tools/clang/include/clang/Driver/Distro.h
@@ -28,6 +28,7 @@ public:
     // the first and last known member in the family, e.g. IsRedHat().
     AlpineLinux,
     ArchLinux,
+    ClearLinux,
     DebianLenny,
     DebianSqueeze,
     DebianWheezy,
@@ -131,6 +132,9 @@ public:
   bool IsGentoo() const {
     return DistroVal == Gentoo;
   }
+  bool IsClearLinux() const {
+    return DistroVal == ClearLinux;
+  }
 
   /// @}
 };
diff --git llvm-project/tools/clang/lib/Driver/Distro.cpp llvm-project/tools/clang/lib/Driver/Distro.cpp
index 4d58ad1ae78c..b0f33398fc8e 100644
--- llvm-project/tools/clang/lib/Driver/Distro.cpp
+++ llvm-project/tools/clang/lib/Driver/Distro.cpp
@@ -165,6 +165,19 @@ static Distro::DistroType DetectDistro(llvm::vfs::FileSystem &VFS,
   if (VFS.exists("/etc/gentoo-release"))
     return Distro::Gentoo;
 
+  File = VFS.getBufferForFile("/usr/lib/os-release");
+  if (File) {
+    StringRef Data = File.get()->getBuffer();
+    SmallVector<StringRef, 16> Lines;
+    Data.split(Lines, "\n");
+    Distro::DistroType Version = Distro::UnknownDistro;
+    for (StringRef Line : Lines)
+      if (Version == Distro::UnknownDistro && Line.startswith("ID_LIKE="))
+        Version = llvm::StringSwitch<Distro::DistroType>(Line.substr(7))
+                      .Case("clear-linux-os", Distro::ClearLinux);
+    return Version;
+  }
+
   return Distro::UnknownDistro;
 }
 
diff --git llvm-project/tools/clang/lib/Driver/ToolChains/Linux.cpp llvm-project/tools/clang/lib/Driver/ToolChains/Linux.cpp
index 180350476c38..dce87427bdbb 100644
--- llvm-project/tools/clang/lib/Driver/ToolChains/Linux.cpp
+++ llvm-project/tools/clang/lib/Driver/ToolChains/Linux.cpp
@@ -221,13 +221,13 @@ Linux::Linux(const Driver &D, const llvm::Triple &Triple, const ArgList &Args)
 
   Distro Distro(D.getVFS(), Triple);
 
-  if (Distro.IsAlpineLinux() || Triple.isAndroid()) {
+  if (Distro.IsAlpineLinux() || Distro.IsClearLinux() || Triple.isAndroid()) {
     ExtraOpts.push_back("-z");
     ExtraOpts.push_back("now");
   }
 
   if (Distro.IsOpenSUSE() || Distro.IsUbuntu() || Distro.IsAlpineLinux() ||
-      Triple.isAndroid()) {
+      Distro.IsClearLinux() || Triple.isAndroid()) {
     ExtraOpts.push_back("-z");
     ExtraOpts.push_back("relro");
   }
@@ -282,7 +282,7 @@ Linux::Linux(const Driver &D, const llvm::Triple &Triple, const ArgList &Args)
   ExtraOpts.push_back("--build-id");
 #endif
 
-  if (IsAndroid || Distro.IsOpenSUSE())
+  if (IsAndroid || Distro.IsOpenSUSE() || Distro.IsClearLinux())
     ExtraOpts.push_back("--enable-new-dtags");
 
   // The selection of paths to try here is designed to match the patterns which
@@ -819,8 +819,12 @@ void Linux::AddIAMCUIncludeArgs(const ArgList &DriverArgs,
 }
 
 bool Linux::isPIEDefault() const {
-  return (getTriple().isAndroid() && !getTriple().isAndroidVersionLT(16)) ||
-          getTriple().isMusl() || getSanitizerArgs().requiresPIE();
+  if ((getTriple().isAndroid() && !getTriple().isAndroidVersionLT(16)) ||
+      getTriple().isMusl() || getSanitizerArgs().requiresPIE())
+    return true;
+
+  Distro Distro(getDriver().getVFS(), getTriple());
+  return Distro.IsClearLinux();
 }
 
 bool Linux::isNoExecStackDefault() const {
diff --git llvm-project/tools/clang/unittests/Driver/DistroTest.cpp llvm-project/tools/clang/unittests/Driver/DistroTest.cpp
index 7c00a8f66be8..318833938f7c 100644
--- llvm-project/tools/clang/unittests/Driver/DistroTest.cpp
+++ llvm-project/tools/clang/unittests/Driver/DistroTest.cpp
@@ -342,6 +342,22 @@ TEST(DistroTest, DetectGentoo) {
   ASSERT_TRUE(Gentoo.IsGentoo());
 }
 
+TEST(DistroTest, DetectClearLinux) {
+  vfs::InMemoryFileSystem ClearLinuxFileSystem;
+  ClearLinuxFileSystem.addFile("/usr/lib/os-release", 0,
+      llvm::MemoryBuffer::getMemBuffer("NAME=\"Clear Linux OS\"\n"
+                                       "VERSION=1\n"
+                                       "ID=clear-linux-os\n"
+                                       "ID_LIKE=clear-linux-os\n"
+                                       "VERSION_ID=25530\n"
+                                       "PRETTY_NAME=\"Clear Linux OS\"\n"
+                                       "ANSI_COLOR=\"1;35\"\n"
+                                       "HOME_URL=\"https://clearlinux.org\"\n"
+                                       "SUPPORT_URL=\"https://clearlinux.org\"\n"
+                                       "BUG_REPORT_URL=\"mailto:dev@lists.clearlinux.org\"\n"
+                                       "PRIVACY_POLICY_URL=\"http://www.intel.com/privacy\"\n"));
+}
+
 TEST(DistroTest, DetectWindowsAndCrossCompile) {
 
   class CountingFileSystem : public llvm::vfs::ProxyFileSystem {
-- 
2.31.1

